### PR TITLE
feat: Add Notebook Controller

### PR DESCRIPTION
**Pull request by wg102**
Closes https://github.com/StatCan/daaas/issues/546

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+var (
+	masterURL      string
+	kubeconfig     string
+)
+
+func main() {
+
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
```
Comment (justbert) [inline main.go]: These can go in an `init()` function:
https://tutorialedge.net/golang/the-go-init-function/

File: main.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		istioclient,
+		kubeflowclient,
+		kubeflowInformerFactory.Kubeflow().V1().Notebooks(),
+		istioInformerFactory.Security().V1beta1().AuthorizationPolicies(),
+	)
+
+	kubeInformerFactory.Start(stopCh)
+	istioInformerFactory.Start(stopCh)
+	kubeflowInformerFactory.Start(stopCh)
+
```
Comment (justbert) [inline main.go:68]: I'm not sure if this will work. I believe only one of these will be able to receive the cancel signal.

**Review by justbert — APPROVED** [review]
This looks good to me! Just take a look at the small comment I have!

@zachomedia and I have been talking about e2e (end-to-end) testing frameworks so we'll keep you in the loop!

You might want to check in on statcan/profiles-controller! @zachomedia built it in a way that removes a lot of boilerplate.

File: main.go (hunk 1 above)
Comment (wg102) [reply main.go]: Managed the flag in `init()` function. Let me know if there was more for this

File: main.go (hunk 2 above)
Comment (zachomedia) [reply main.go:68]: Ah! I finally figured out why this works! It actually closes the channel which will be forwarded to everyone.

https://github.com/StatCan/prob-notebook-controller/pull/1/files#diff-21b08173ead80fa0502f2688fbef35144ebe257cadb326fef9355cf0201b0716R37

File: main.go (hunk 1 above)
Comment (justbert) [reply main.go]: It's good! A best practice is to run the `flag.Parse()` in the `main`. This ensures that all other modules that are setting flags via an `init()` in their pacakages will also be parsed.

`init()` functions are run before `main`, however, I don't believe the order in which they are run is well defined nor consistent. As such, it's possible that the `flag.Parse()` function is run before all the flags are setup and could cause them not to be initialized correctly.

File: main.go (hunk 2 above)
Comment (justbert) [reply main.go:68]: Aha! Oh right!

**Pull request merged**